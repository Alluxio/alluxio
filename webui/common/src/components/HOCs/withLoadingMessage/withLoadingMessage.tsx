/*
 * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
 * (the "License"). You may not use this work except in compliance with the License, which is
 * available at www.apache.org/licenses/LICENSE-2.0
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied, as more fully set forth in the License.
 *
 * See the NOTICE file distributed with this work for information regarding copyright ownership.
 */

import React from 'react';
import {getDisplayName} from "../../../utilities";
import {LoadingMessage} from "../..";

export interface ILoadingProps {
    loading: boolean;
    class: string;
}

export function withLoadingMessage<T extends ILoadingProps>(WrappedComponent: React.ComponentType<T>) {
    const loaderHoc = (props: T) => {
        return props.loading
            ? (
                <div className={props.class}>
                    <LoadingMessage />
                </div>
            )
            : <WrappedComponent {...props} />;
    };
    (loaderHoc as React.FunctionComponent).displayName = `withLoadingMessage(${getDisplayName(WrappedComponent)})`;
    return loaderHoc;
}
