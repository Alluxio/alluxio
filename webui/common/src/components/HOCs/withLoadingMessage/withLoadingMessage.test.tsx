/*
 * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
 * (the "License"). You may not use this work except in compliance with the License, which is
 * available at www.apache.org/licenses/LICENSE-2.0
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied, as more fully set forth in the License.
 *
 * See the NOTICE file distributed with this work for information regarding copyright ownership.
 */

import {configure, shallow, ShallowWrapper} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import {ILoadingProps, withLoadingMessage} from "./withLoadingMessage";

configure({adapter: new Adapter()});

const WrappedComponent = () => <div>Wrapped</div>;
const EnhancedComponent = withLoadingMessage(WrappedComponent);

describe('withLoadingMessage HOC', () => {
    let props: ILoadingProps;

    beforeAll(() => {
        props = {
            loading: false,
            class: 'test'
        };
    });

    describe('Shallow component', () => {
        let shallowWrapper: ShallowWrapper;

        beforeAll(() => {
            shallowWrapper = shallow(<EnhancedComponent {...props}/>);
        });

        it('Renders without crashing', () => {
            expect(shallowWrapper.length).toEqual(1);
        });

        describe('Loading', () => {
            beforeAll(() => {
                shallowWrapper.setProps({ loading: true });
            });

            it('Contains a div with class test', () => {
                expect(shallowWrapper.find('.test').length).toEqual(1);
            });

            it('Matches snapshot - renders LoadingMessage', () => {
                expect(shallowWrapper).toMatchSnapshot();
            });
        });

        describe('Not Loading', () => {
            it('Matches snapshot - renders WrappedComponent', () => {
                shallowWrapper.setProps({ loading: false });
                expect(shallowWrapper).toMatchSnapshot();
            });
        });
    });
});
