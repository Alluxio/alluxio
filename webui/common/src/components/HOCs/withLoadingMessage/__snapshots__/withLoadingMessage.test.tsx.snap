// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withLoadingMessage HOC Shallow component Loading Matches snapshot - renders LoadingMessage 1`] = `
<div
  className="test"
>
  <LoadingMessage />
</div>
`;

exports[`withLoadingMessage HOC Shallow component Not Loading Matches snapshot - renders WrappedComponent 1`] = `
<WrappedComponent
  class="test"
  loading={false}
/>
`;
